File: apps/api/src/routers/observations.py
import asyncpg
from fastapi import APIRouter, Depends, HTTPException, Query, Request

from config import limiter
from deps import get_pool, require_admin
from observations import store
from observations.api_models import (
    ObservedFactCreateRequest,
    ObservedFactDeleteResponse,
    ObservedFactListResponse,
    ObservedFactResponse,
    ObservedFactUpdateRequest,
    ObservationFactSummaryResponse,
    PredictionObservationCompareRequest,
    PredictionObservationCompareResponse,
    ValidationReviewRequest,
    ValidationReviewResponse,
)
# Stage 6C: import the comparison engine directly from its submodule so we
# do NOT accidentally pick up the Stage 4D ``compare_prediction_to_observations``
# re-exported by ``observations/__init__.py``. The Stage 6C engine has a
# different signature and a different result shape.
from observations.comparison_engine import (
    compare_prediction_to_observations as compare_prediction_to_observations_stage6c,
)
from observations.comparison_models import comparison_result_to_dict
from observations.review_engine import build_validation_review
from observations.review_models import review_result_to_dict

router = APIRouter(tags=['observations'])

_OPERATOR_MUTATION_LIMIT = '20/minute'


async def _load_or_convert_comparison_facts(
    *,
    body: PredictionObservationCompareRequest,
    pool: asyncpg.Pool,
):
    """Shared Stage 6C/6E fact-loading semantics."""
    if body.observed_facts is None:
        facts, _total = await store.list_observed_facts_for_comparison(
            pool,
            system_id64=body.system_id64,
            target_archetype=body.target_archetype,
            limit=body.fact_load_limit,
            offset=0,
        )
        return list(facts)
    return [fact_input.to_persisted() for fact_input in body.observed_facts]


@router.post(
    '/api/observations/facts',
    response_model=ObservedFactResponse,
    dependencies=[Depends(require_admin)],
)
@limiter.limit(_OPERATOR_MUTATION_LIMIT)
async def create_observed_fact(
    request: Request,
    body: ObservedFactCreateRequest,
    pool: asyncpg.Pool = Depends(get_pool),
) -> ObservedFactResponse:
    fact = await store.create_observed_fact(pool, body)
    return ObservedFactResponse.from_domain(fact)


@router.get('/api/observations/facts', response_model=ObservedFactListResponse)
async def list_observed_facts(
    system_id64: int = Query(..., gt=0),
    fact_type: str | None = None,
    subject_type: str | None = None,
    status: str | None = None,
    target_archetype: str | None = None,
    build_fingerprint: str | None = None,
    simulation_fingerprint: str | None = None,
    limit: int = Query(100, ge=1, le=500),
    offset: int = Query(0, ge=0),
    pool: asyncpg.Pool = Depends(get_pool),
) -> ObservedFactListResponse:
    facts, total = await store.list_observed_facts(
        pool,
        system_id64=system_id64,
        fact_type=fact_type,
        subject_type=subject_type,
        status=status,
        target_archetype=target_archetype,
        build_fingerprint=build_fingerprint,
        simulation_fingerprint=simulation_fingerprint,
        limit=limit,
        offset=offset,
    )
    # Summary must describe the full filtered result set, not just the
    # paginated page returned above. We re-run the same filter without
    # limit/offset so total and summary describe identical row sets.
    summary = await store.summarise_observed_facts_for_filter(
        pool,
        system_id64=system_id64,
        fact_type=fact_type,
        subject_type=subject_type,
        status=status,
        target_archetype=target_archetype,
        build_fingerprint=build_fingerprint,
        simulation_fingerprint=simulation_fingerprint,
    )
    return ObservedFactListResponse(
        facts=[ObservedFactResponse.from_domain(fact) for fact in facts],
        total=total,
        limit=limit,
        offset=offset,
        summary=ObservationFactSummaryResponse.from_domain(summary),
    )


@router.get('/api/observations/facts/{observation_id}', response_model=ObservedFactResponse)
async def get_observed_fact(
    observation_id: str,
    pool: asyncpg.Pool = Depends(get_pool),
) -> ObservedFactResponse:
    fact = await store.get_observed_fact(pool, observation_id)
    if fact is None:
        raise HTTPException(404, f'Observed fact {observation_id} not found')
    return ObservedFactResponse.from_domain(fact)


@router.patch(
    '/api/observations/facts/{observation_id}',
    response_model=ObservedFactResponse,
    dependencies=[Depends(require_admin)],
)
@limiter.limit(_OPERATOR_MUTATION_LIMIT)
async def update_observed_fact(
    request: Request,
    observation_id: str,
    body: ObservedFactUpdateRequest,
    pool: asyncpg.Pool = Depends(get_pool),
) -> ObservedFactResponse:
    fact = await store.update_observed_fact(pool, observation_id, body)
    if fact is None:
        raise HTTPException(404, f'Observed fact {observation_id} not found')
    return ObservedFactResponse.from_domain(fact)


@router.delete(
    '/api/observations/facts/{observation_id}',
    response_model=ObservedFactDeleteResponse,
    dependencies=[Depends(require_admin)],
)
@limiter.limit(_OPERATOR_MUTATION_LIMIT)
async def delete_observed_fact(
    request: Request,
    observation_id: str,
    pool: asyncpg.Pool = Depends(get_pool),
) -> ObservedFactDeleteResponse:
    deleted = await store.delete_observed_fact(pool, observation_id)
    if not deleted:
        raise HTTPException(404, f'Observed fact {observation_id} not found')
    return ObservedFactDeleteResponse(observation_id=observation_id, deleted=True)


# ──────────────────────────────────────────────────────────────────────
# Stage 6C — predicted-vs-observed comparison endpoint
# ──────────────────────────────────────────────────────────────────────
@router.post(
    '/api/observations/compare',
    response_model=PredictionObservationCompareResponse,
)
async def compare_prediction_against_observations(
    body: PredictionObservationCompareRequest,
    pool: asyncpg.Pool = Depends(get_pool),
) -> PredictionObservationCompareResponse:
    """Run the Stage 6C deterministic comparison engine.

    Two modes:

    * **Mode A** — ``observed_facts`` omitted from the request. The
      router loads up to ``fact_load_limit`` persisted facts for
      ``system_id64`` via ``store.list_observed_facts_for_comparison``
      and passes them to the engine.

      Fact-loading semantics (Stage 6C hardening):
        * If ``target_archetype`` is omitted/null in the request, all
          facts for the system are loaded.
        * If ``target_archetype`` is supplied, facts are included when
          their ``target_archetype`` matches the requested value **or**
          is ``NULL`` — i.e. system-level general evidence (notes,
          service evidence, economy evidence, CP observations not tied
          to a specific target) is included alongside target-specific
          evidence. Facts targeting a *different* explicit archetype
          are excluded.
    * **Mode B** — ``observed_facts`` provided. The supplied list is
      converted to ``PersistedObservedFact`` objects and passed to the
      engine verbatim. The database is NOT queried for facts in this
      mode.

      Source semantics (Stage 6C hardening): Mode B supplied facts may
      carry any ``ObservationSource`` value including ``imported`` and
      ``inferred``. This is deliberate — Mode B is read-only and the
      supplied facts are never persisted, so Stage 6A's reserved-source
      restriction (manual/test_fixture only on create/update) does not
      apply. Persistence still goes exclusively through Stage 6A's
      ``POST /api/observations/facts`` endpoint and remains restricted.

    The handler is **passive**: it never invokes simulation, optimiser,
    or ranking code, and never mutates persisted observations. It only
    reads observations (Mode A) and runs the pure comparison engine.
    """
    observed = await _load_or_convert_comparison_facts(body=body, pool=pool)

    result = compare_prediction_to_observations_stage6c(
        system_id64=body.system_id64,
        target_archetype=body.target_archetype,
        prediction=body.prediction,
        observed_facts=observed,
    )

    # ``comparison_result_to_dict`` produces JSON-safe primitives that
    # round-trip directly into the response Pydantic model.
    return PredictionObservationCompareResponse.model_validate(
        comparison_result_to_dict(result),
    )


@router.post(
    '/api/observations/review',
    response_model=ValidationReviewResponse,
)
async def review_prediction_validation(
    body: ValidationReviewRequest,
    pool: asyncpg.Pool = Depends(get_pool),
) -> ValidationReviewResponse:
    """Run Stage 6C comparison and Stage 6E review guidance.

    The handler mirrors compare-endpoint input semantics for caller
    convenience, then passes the comparison result into the pure Stage
    6E review engine. It does not run Simulation Preview, optimiser
    generation, optimiser ranking, or any mechanics module, and it does
    not mutate observations or predictions.
    """
    observed = await _load_or_convert_comparison_facts(body=body, pool=pool)
    comparison_result = compare_prediction_to_observations_stage6c(
        system_id64=body.system_id64,
        target_archetype=body.target_archetype,
        prediction=body.prediction,
        observed_facts=observed,
    )
    review_result = build_validation_review(comparison_result=comparison_result)
    return ValidationReviewResponse.model_validate(review_result_to_dict(review_result))

File: frontend/src/lib/api/observations.ts
import type {
  ListObservedFactsParams,
  ObservedFact,
  ObservedFactCreateRequest,
  ObservedFactDeleteResponse,
  ObservedFactListResponse,
  ObservedFactUpdateRequest,
  PredictionObservationCompareRequest,
  PredictionObservationCompareResponse,
  ValidationReviewRequest,
  ValidationReviewResponse,
} from '@/types/api';
import { jsonFetch, operatorMutationHeaders } from './core';

// ── Stage 6B Observed Evidence (Observed Facts) ────────────────────────
// Passive evidence: records what a user actually saw in-game for a system
// or build. These calls do NOT change Simulation Preview scoring, optimiser
// ranking, or generated candidates. They are also not consumed by the
// simulation/optimiser modules — see api-contracts.md "Stage 6A Observed
// Facts API" for the contract details.
export function listObservedFacts(params: ListObservedFactsParams): Promise<ObservedFactListResponse> {
  const usp = new URLSearchParams();
  usp.set('system_id64', String(params.system_id64));
  if (params.fact_type)              usp.set('fact_type',              params.fact_type);
  if (params.subject_type)           usp.set('subject_type',           params.subject_type);
  if (params.status)                 usp.set('status',                 params.status);
  if (params.target_archetype)       usp.set('target_archetype',       params.target_archetype);
  if (params.build_fingerprint)      usp.set('build_fingerprint',      params.build_fingerprint);
  if (params.simulation_fingerprint) usp.set('simulation_fingerprint', params.simulation_fingerprint);
  if (params.limit  !== undefined)   usp.set('limit',  String(params.limit));
  if (params.offset !== undefined)   usp.set('offset', String(params.offset));
  return jsonFetch(`/observations/facts?${usp.toString()}`);
}

export function createObservedFact(request: ObservedFactCreateRequest): Promise<ObservedFact> {
  return jsonFetch('/observations/facts', {
    method: 'POST',
    body:   JSON.stringify(request),
    headers: operatorMutationHeaders(),
  });
}

export function updateObservedFact(observationId: string, request: ObservedFactUpdateRequest): Promise<ObservedFact> {
  return jsonFetch(`/observations/facts/${encodeURIComponent(observationId)}`, {
    method: 'PATCH',
    body:   JSON.stringify(request),
    headers: operatorMutationHeaders(),
  });
}

export function deleteObservedFact(observationId: string): Promise<ObservedFactDeleteResponse> {
  return jsonFetch(`/observations/facts/${encodeURIComponent(observationId)}`, {
    method: 'DELETE',
    headers: operatorMutationHeaders(),
  });
}

// ── Stage 6C Predicted-vs-Observed Comparison ──────────────────────────
// Read-only comparison: takes a current prediction (a
// SimulateBuildResponse, in practice) and asks the backend to compare
// it against persisted observed evidence for the same system. Stage 6D
// renders the result inside Colony Planner. This call does NOT mutate
// any prediction, optimiser candidate, optimiser ranking, or persisted
// observation. The backend operates in Mode A by default (it loads
// persisted facts itself); Stage 6D never sends `observed_facts`.
export function comparePredictionToObservations(
  request: PredictionObservationCompareRequest,
): Promise<PredictionObservationCompareResponse> {
  return jsonFetch('/observations/compare', {
    method: 'POST',
    body:   JSON.stringify(request),
  });
}

// ── Stage 6E Validation Review Guidance ────────────────────────────────
// Read-only advisory guidance built from the Stage 6C comparison
// result. This helper only calls the review endpoint; it does not run
// Simulation Preview, optimiser candidate generation, or observation
// mutations.
export function reviewPredictionValidation(
  request: ValidationReviewRequest,
): Promise<ValidationReviewResponse> {
  return jsonFetch('/observations/review', {
    method: 'POST',
    body:   JSON.stringify(request),
  });
}
